From 1c8aef678d8de7d3dda8c5f67ab696f5ecacfe2a Mon Sep 17 00:00:00 2001
From: Arie Brown <abrown@habana.ai>
Date: Sun, 12 Jul 2020 09:59:00 +0300
Subject: [PATCH] [SW-13943] fit: read pubkey from apart image

the public key was saved as part of u-boot binary image and it was
change.
the public key stored as a separate image and in that way we can use
it for several configurations.

Change-Id: I1d98a2bf7bc068f560a91d669aff0cae3bebca0d
Signed-off-by: Arie Brown <abrown@habana.ai>
---

diff --git a/preboot/fit.c b/preboot/fit.c
index edd3804..9552ff8 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -26,7 +26,7 @@
 #define FIT_VALUE_PROP		"value"
 #define FIT_ALGO_PROP		"algo"
 #define FIT_KEY_NAME_HINT	"key-name-hint"
-#define FIT_HL_PUBKEY_LEN_PROP  "hl-pub-key-len"
+#define FIT_PUB_KEY_PROP	"pub-key"
 
 #define FIT_ALGO_PROP_EXPECTED_VALUE	"sha256,rsa3072"
 
@@ -38,6 +38,13 @@
 static const char *fit_loadables_name(const void *fit, int noffset, int id)
 {
 	return fdt_stringlist_get(fit, noffset, FIT_LOADABLE_PROP, id, NULL);
+}
+
+static const char *fit_pub_key_name(const void *fit, int noffset)
+{
+	int prop_len;
+
+	return fdt_getprop(fit, noffset, FIT_PUB_KEY_PROP, &prop_len);
 }
 
 static int fit_find_config_node(const void *fit)
@@ -207,12 +214,11 @@
  * from u-boot: u-boot/common/image-sig.c
  */
 int fit_check_sig(const void *fit, int noffset, const u8 *data, int data_len,
-		  int pub_key_len)
+		  const void *pub_key_fit)
 {
 	const u8 *algo_prop = NULL;
 	const u8 *hash_value = NULL;
 	const char *key_name = NULL;
-	const void *pub_key_fit = NULL;
 	int hash_len = 0;
 	struct rsa_key_prop prop;
 
@@ -225,18 +231,50 @@
 	if (get_key_name(fit, noffset, &key_name))
 		return 1;
 
-	pub_key_fit = data + data_len - pub_key_len;
 	if (fit_get_key_prop(pub_key_fit, key_name, &prop))
 		return 1;
 
 	return rsa_verify(data, data_len, hash_value, hash_len, &prop);
 }
 
-int fit_verify_image(const void *fit, int node, const u8 *buf, int *len)
+static int fit_pub_key_node_get(const void *fit, int conf)
+{
+	int images;
+	const char *uname = fit_pub_key_name(fit, conf);
+
+	images = fdt_path_offset(fit, FIT_IMAGES_PATH);
+	if (uname)
+		return fdt_subnode_offset(fit, images, uname);
+
+	return -1;
+}
+
+static int fit_get_pub_key(const void *fit, int noffset,
+			   const void **pub_key_fit)
+{
+	int pub_key_fit_len;
+	int node;
+
+	node = fit_pub_key_node_get(fit, noffset);
+	if (node < 0) {
+		preboot_puts("Wrong FIT format missing pubkey\n");
+		return 1;
+	}
+
+	*pub_key_fit = fdt_getprop(fit, node, FIT_DATA_PROP, &pub_key_fit_len);
+	if (!(*pub_key_fit)) {
+		preboot_puts("Wrong FIT format: missing pubkey fit\n");
+		return 1;
+	}
+
+	return 0;
+}
+
+int fit_verify_image(const void *fit, int node, const u8 *buf, int buf_len,
+		     int noffset)
 {
 	int sig_sub_node;
-	int pub_key_len;
-	int buf_len = *len;
+	const void *pub_key_fit;
 	int ret = 1;
 
 	sig_sub_node = fdt_subnode_offset(fit, node, FIT_SIG_NODENAME);
@@ -245,19 +283,18 @@
 		return 1;
 	}
 
-	pub_key_len = fit_get_int(fit, 0, FIT_HL_PUBKEY_LEN_PROP, 0);
-	if (pub_key_len == 0) {
-		preboot_puts("Wrong FIT format: public key length invalid\n");
+	if (fit_get_pub_key(fit, noffset, &pub_key_fit)) {
+		preboot_puts("Wrong FIT format: missing pub_key\n");
 		return 1;
 	}
-	if (!fit_check_sig(fit, sig_sub_node, buf, buf_len, pub_key_len)) {
+
+	if (!fit_check_sig(fit, sig_sub_node, buf, buf_len, pub_key_fit)) {
 		preboot_puts("fit_check_sig OK\n");
 		ret = 0;
 	} else {
 		preboot_puts("fit_check_sig failed\n");
 		ret = 1;
 	}
-	*len = *len - pub_key_len;
 	return ret;
 }
 
@@ -271,6 +308,9 @@
 
 	/* Get number of images in preboot configuration */
 	conf = fit_find_config_node(fit);
+	if (conf < 0)
+		return 1;
+
 	num_images = fdt_stringlist_count(fit, conf, FIT_LOADABLE_PROP);
 	if (num_images > FIT_MAX_IMAGES) {
 		preboot_puts_hex("Images in fit exceeds max limit: ",
@@ -308,7 +348,8 @@
 		}
 
 		if (fit_need_image_verification())
-			copy_image = !fit_verify_image(fit, node, buf, &len);
+			copy_image = !fit_verify_image(fit, node, buf, len,
+						       conf);
 
 		images[i].loadaddr = dst;
 
